### PR TITLE
FMS2: Replace time_interp_external IDs with a MOM-defined type

### PR DESCRIPTION
**Pull request by marshallward**
All instances of an FMS ID to the internal interpolation content is replaced with a derived type containing additional metadata recording the field's origin filename and fieldname.

This additional information is required in order to replicate the axis data from the field, which is no longer provided by FMS2.

The abstraction of this type also allows us to either extend it or redefine it in other frameworks as needed in the future.

This primarily affects the usage of the following functions:

- init_external_field
- time_interp_external
- horiz_interp_and_extrap_tracer

References to FMS1 move data into an `axistype`, but the contents are transferred to an `axis_info` when leaving the FMS1 layer.  FMS2 directly populates the `axis_info` content.

The `get_external_field_info` calls are modified to return `axis_info` rather than `axistype`.

The redundant `get_axis_data` function is also removed from `MOM_interp_infra`, since `get_axis_info` provides an equivalent operation.

The following solvers are updated:

- MOM_open_boundary
- MOM_ice_shelf
- MOM_oda_driver
- MOM_MEKE
- MOM_ALE_sponge
- MOM_diabatic_aux

Of these, OBC was the most significant.  The integer handle (fid) was previously used to determine if each segment field was constant or (if negative) read from a file.  After being replaced by the derived type, a new flag was added to make this determination.

All of the coupled drivers have been modified, since they support time interpolation of T and S fields.

- FMS
- MCT
- NUOPC

The NUOPC driver also includes modifications to its CFC11 and CFC12 fields.  Changes to the MOM CFC modules replaces an `id == -1`-like test, which is not used by the derived type.  This check has been removed, and we now solely rely on the `present(cfc_handle)` test.

----

This does not resolve all of our dependencies on FMS1, since content from `time_interp_external` must be replaced with content from `time_interp_external2`.  But this PR allows us to begin the work of using the replacement module.

### CODE REVIEW

**Comment by marshallward** [comment]
Since this touches the NUOPC and MCT couplers, I think it's important to get feedback (if not approval) from NCAR.  @gustavo-marques @alperaltuntas @mnlevy1981 Are any of you able to review this?

Also feedback from @MJHarrison-GFDL @andrew-c-ross @kshedstrom would be appreciated.

Having said that, there will probably be a more intensive PR coming later which replaces the actual `time_interp_external` calls.

**Comment by jiandewang** [comment]
@marshallward since there are changes in NUOPC_cap, I am going to test it in UFS.

**Comment by marshallward** [comment]
Thanks @jiandewang I didn't realize you were using NUOPC.

**Comment by kshedstrom** [comment]
In repro mode, my Supercritical test is hanging after printing:
```
NOTE from PE     0: diag_manager_mod::diag_manager_init: diag_manager is using fms2_io
NOTE from PE     0: diag_manager_mod::diag_manager_init: prepend_date only supported when diag_manager_init is called with time_init present.
 MOM_in domain decomposition
whalo =    4, ehalo =    4, shalo =    4, nhalo =    4
  X-AXIS =  100 100
  Y-AXIS =  150
 MOM_in domain decomposition
whalo =    2, ehalo =    2, shalo =    2, nhalo =    2
  X-AXIS =   50  50
  Y-AXIS =   75
 MOM_MOSAIC domain decomposition
whalo =    9, ehalo =    9, shalo =    9, nhalo =    9
  X-AXIS =  200 200
  Y-AXIS =  300
 MOM_MOSAIC domain decomposition
whalo =    4, ehalo =    4, shalo =    4, nhalo =    4
  X-AXIS =  100 100
  Y-AXIS =  150
```
It runs fine in debug mode, where the next thing printed is:
```
WARNING from PE     0: Open_file is appending .nc to the filename ./ocean_geometry
```
It also runs fine with one core in repro mode. The ones that hang are using 2-4 cores.

**Comment by kshedstrom** [comment]
It runs on gaea with ifort22.
It runs with FMS 2023.01.
It fails on chinook with FMS 2023.01.01 and gfortran 11, but only when using multiple cores.
CCS2 (coupled MOM6-SIS2) also fails on chinook.

**Comment by kshedstrom** [comment]
Never mind, it's an FMS problem, not a problem with this patch.

**Comment by klindsay28** [comment]
I migrated a lot of the CFC_cap code out of the NUOPC driver and into MOM_tracer_flow_control.F90 and MOM_CFC_cap.F90 with https://github.com/NCAR/MOM6/pull/242. So if dev/gfdl gets to mom-ocean/main before dev/ncar, we'll need to merge these mods into dev/ncar, or you'll get to do it if dev/ncar gets to mom-ocean/main before dev/gfdl.

**Comment by marshallward** [comment]
@klindsay28 Enough problems are starting to bubble up that I would not expect this in the next PR to main, so most likely we'll have to wrangle with the merge conflicts.  Thanks for the heads up though.

**Comment by marshallward** [comment]
I updated this PR to replace the `time_interp_external` module calls with `time_interp_external2`.  This includes a change which assumes that time-interpolated external fields are always on-grid (via `ongrid=.true.` in `init_external_field`.  (Thanks to @uramirez8707 for the suggestion.)

This appears to be a safe assumption, but I could be wrong about it.  But if we assume it, then it resolves the few outstanding issues with adopting `time_interp_external2`.

**Comment by marshallward** [comment]
Another issue has emerged: It seems that the FMS1 `init_external_field` was case insensitive, but the new FMS2 version is case-sensitive.  I am looking into how to handle this.

**Comment by marshallward** [comment]
I added a commit to support case-insensitive field access in FMS2, which appears to work in my local testing.  As it is, I believe this particular PR is ready to merge.

However, it might be helpful to resolve the (presumably unrelated) issue detected by @jiandewang before merging, so that it can be tested in UFS.

**Comment by kshedstrom** [comment]
The current version of this runs the dumbbell default tests, the dumbbell scaled tests, the dumbbell rotated test, but not the dumbbell scaled and rotated test. Stack trace:
```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x7f18ec73856f in ???
#1  0x11d187d in __mom_interpolate_MOD_time_interp_external_3d
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/framework/MOM_interpolate.F90:174
#2  0x4cd5b3 in __mom_open_boundary_MOD_update_obc_segment_data
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM_open_boundary.F90:3898
#3  0x15ed6c2 in __mom_boundary_update_MOD_update_obc_data
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM_boundary_update.F90:160
#4  0xf37547 in __mom_dynamics_split_rk2_MOD_step_mom_dyn_split_rk2
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM_dynamics_split_RK2.F90:507
#5  0x18d52e6 in step_mom_dynamics
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM.F90:1174
#6  0x18de1f1 in __mom_MOD_step_mom
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM.F90:853
#7  0x151d7b4 in mom6
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/config_src/drivers/solo_driver/MOM_driver.F90:484
#8  0x151ed0f in main
	at //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/config_src/drivers/solo_driver/MOM_driver.F90:27
```
Something isn't right with the rotation of data.in.

**Comment by marshallward** [comment]
@kshedstrom Are you saying this fails in `dev/gfdl` as well?  Or just in this PR?

**Comment by kshedstrom** [comment]
@marshallward Sorry, answers don't change from dev/gfdl. It's just that I was in debugger mode to look at something else and it tripped over the invalid numbers. But I do need to figure out what's wrong with my rotated cases now that I have the debug executable.

**Comment by kshedstrom** [comment]
Running on chinook04 (gcc 11.3), answers match between the two MOM6 versions. Going back to chinook01 (gcc 8.3) with dev/gfdl, the answers still mostly match in repro mode, also in debug mode without this flag:  -ffpe-trap=invalid,zero,overflow
Turning it back on again, I managed to reproduce my blow-up. I need to ask about a debugger on chinook04. Sorry to have bothered you with a compiler bug.

File: config_src/infra/FMS2/MOM_interp_infra.F90
Comment (Hallberg-NOAA) [inline config_src/infra/FMS2/MOM_interp_infra.F90:12]: These module use statements for routines in MOM_io, MOM_error_handler and MOM_string_functions introduce a backward dependency on code in the src/framework directory and the potential for a predecessor cycle.  There are no other instances of config_src/infra code on src/framework code, and I think that we should keep it that way.  Please consider revising this dependency.

File: config_src/infra/FMS1/MOM_interp_infra.F90
Comment (Hallberg-NOAA) [inline config_src/infra/FMS1/MOM_interp_infra.F90:7]: These module use statements for MOM_io introduce a backward dependency on code in the src/framework directory and the potential for a predecessor cycle.  There are no other instances of config_src/infra code on src/framework code, and I think that we should keep it that way.  Please consider revising this dependency.

**Review by Hallberg-NOAA — DISMISSED** [review]
Although I agree with the broad direction of this PR, it has introduced (for the first time) the use of code from the `src/framework` directory by code in the `config_src/infra` directories.  I think that this is potentially dangerous, in that it could lead to predecessor cycles, and that it could inhibit the compilation of parts of MOM6 as libraries.  I would like to see this PR revised to avoid this backward dependency, or else I would like to have a broader discussion about the relative roles of the code in the various `src/` and `config_src/` directories before we collectively adopt this substantial revision to the effective code structure of the model.

**Comment by marshallward** [comment]
`MOM_error` is easily replaced with `MOM_err`, and maybe something can be done about `axis_info` (perhaps the infra-specific `axistype` can be used in its place, although IIRC the problem is that there is no analogue to `set_axis_info`.).  But aside from reimplementing `lowercase()` at the infra level, for the sole purpose of avoiding the appearance of a circular dependency, I can't see any other solution.

IMHO, I see no problem with `config_src` referencing content in `src` when the content is properly platform-independent, as in the case of `lowercase()` and `axis_info`/`set_axis_info`.  But I'll do whatever you request in this case, though I will need some advice on how to move forward.

File: config_src/infra/FMS1/MOM_interp_infra.F90
Comment (Hallberg-NOAA) [reply config_src/infra/FMS1/MOM_interp_infra.F90:7]: This use statement has not created a predecessor cycle between modules.

File: config_src/infra/FMS2/MOM_interp_infra.F90
Comment (Hallberg-NOAA) [reply config_src/infra/FMS2/MOM_interp_infra.F90:12]: These use statements have not created any predecessor cycles between modules.

**Review by Hallberg-NOAA — APPROVED** [review]
These changes are moving the code in the right direction and addressing important problems that have arisen between SIS2 and some versions of FMS.  It should be merged into dev/gfdl now that the pipeline regression testing has passed at gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/19606 .

**Pull request merged**
